include/RAJA/policy/openmp/target_reduce.hpp

```diff
@@ -117,8 +117,8 @@ struct Reduce_Data {
   explicit Reduce_Data(T defaultValue, T identityValue, Offload_Info &info)
       : value(identityValue),
         device{reinterpret_cast<T *>(
-            omp_target_alloc(Teams * sizeof(T), info.deviceID))},
-        host{new T[Teams]}
+            omp_target_alloc(Threads * sizeof(T), info.deviceID))},
```

Comment:
That is true, and that is still what it is actually doing.
Maybe we should add a constexpr MaxNumTeams = ThreadsPerTeam to the reducer to make it more clear this is happening.

Reply:
@MrBurmark I think that would help when reading the code (at least for me!). @rchen20 please do that.

Reply:
I'm a bit confused about the reasoning for something like "constexpr MaxNumTeams = ThreadsPerTeam" in the reducer. Oddly enough, the reducer never specifies or demands the number of teams, so this macro would be compiled away. The reducer relies solely on the number of teams calculated in the execution policy.

If we want a real check to occur, I could try one of the following:
A. In the reducer, add an assert( omp_get_num_teams() <= ThreadsPerTeam ) before the omp pragma in ~TargetReduce().

B. Create a NumTeams member variable in the execution policy, and do a similar sanity check by accessing that member variable in each of the specializations (i.e. ReduceSum, ReduceMin, etc.).

Option A would put the sanity check closest to the potential bug culprit in the reducer, but it would still be relatively unclear who is setting the number of teams. Option B would be fairly clear, but every future specialization will need to implement similar checks, leaving us susceptible to an occasional "gotcha". We could also do both and have overkill on this problem. I'm open to any suggestions.

Reply:
Adding "constexpr MaxNumTeams = ThreadsPerTeam" is intended to be a change in name only. Hopefully the name change will avoid confusion by having the name match its usage in the reducer. This change is also simpler than re-implementing how the size of the allocation is decided/communicated, but doesn't solve the issue you raised about a real check.

To restate the problem, the ThreadsPerTeam check added to the forall is a band-aid over the lack of communication between the reducers and the forall. It relies on the reduction and execution policies having the same values of ThreadsPerTeam. This assumption is not checked and can cause out of bounds accesses when the ThreadsPerTeam in the reduction policy is smaller than the ThreadsPerTeam in the execution policy.

There are a couple of ways to deal with these issues that trade off performance, complexity, and capability.

A. Test inside the loop with omp_get_num_teams. The omp target reducers don't have ideal performance anyway, so adding an assert in their destructors may not be too bad. We would have to try it and see the performance impact.

B. Test outside the loop. Unfortunately the execution policy and the reducers can only communicate through an intermediary. For example, the cuda backend fixes essentially the same problem by communicating the number of blocks through threadlocal global variables. The threadlocal global variables are set before the loop and are then read in the copy constructors of the reducers which can then allocate the correct amount of memory. In this case no cap on the number of blocks is required because the number of blocks is communicated to the reducers.

C. Remove the ThreadsPerTeam template parameter from the reduction policies and use an arbitrary global value for MaxNumTeams. Then MaxNumTeams can be enforced in forall without needing to check for a consistent value in the reducers. This does arbitrarily limit MaxNumTeams, but it was already limited by ThreadsPerTeam.

Reply:
@rchen20 @MrBurmark since we removed the block size parameter from the CUDA reduction policies, we should not have it in the omp target reduction policies for consistency.

Reply:
@MrBurmark Thanks, the reasoning for constexpr makes sense now. Your A and B suggestions are pretty similar to mine, but stated much more clearly. I considered C as well, but didn't think that would be flexible enough.

@rhornung67 @MrBurmark I'll implement the constexpr for clarity, along with the reducer asserts that Jason and I suggested (option As).